base project no tests or stuff

BadJukeBox opened the pull request.

BadJukeBox committed:
base project no tests or stuff

tttruong701 commented on src/main.py:
```diff
@@ -0,0 +1,36 @@
+
+
+def main(args):
+    alert_timer = 0  # every 12 10 second waits we'll also check for the 2m mark
```
Can you clarify this comment? Maybe move this comment above main(), like a method comment?

tttruong701 commented on src/statistics_management.py:
```diff
@@ -0,0 +1,181 @@
+
+class StatisticsManager:
+    """
+    A Statistics computation/accumulation class. contains methods both to calculate statistics, and to display them.
```
Maybe `StatisticsManager aggregates, computes, and displays (curl/traffic/xyz) statistics.)

tttruong701 commented on src/statistics_management.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        :param http_request_list: (list<HTTPRequestPacket>) the request packets gathered in the last 10 seconds.
+        :return: None
+        Checks the request path, and separates it into only the initial section (the first piece of the bath before the
```
"the first piece of the path"

tttruong701 commented on src/statistics_management.py:
```diff
@@ -0,0 +1,181 @@
+            if request.host not in self.sites_recent:
+                self.sites_recent[request.host] = {
+                    'sections': [section],
+                    'site_hit_count': 1
```
'sections' and 'site_hit_count' could be set to constant string variables.

tttruong701 commented on src/main.py:
```diff
@@ -0,0 +1,36 @@
+    while True:
+        sleep(10)
+        alert_timer += 1
+        statistics.calculate_statistics(packet_collector.collection)
+        statistics.show_all_statistics()
+        statistics.clear_recent_statistics()
```
Can this be moved into a single method in StatisticsManager since they're all related?

If these steps change, then the change is localized.

tttruong701 commented on src/main.py:
```diff
@@ -0,0 +1,36 @@
+from collection import HTTPPacketCollector
```
Can you add a README?

tttruong701 reviewed (changes requested):
Still reviewing, some initial comments.

BadJukeBox committed:
Added a README.md, added documentation.
Changed the alert average threshold check to be a float so that alert checking can be tigher.
Got rid of some magic numbers

BadJukeBox committed:
Changed alert_timer to interval_count for name/use clarity

BadJukeBox committed:
Added some basic logging and updated the README.md

tttruong701 commented on src/utils.py:
```diff
@@ -0,0 +1,24 @@
+    err_handler = logging.StreamHandler()
+    err_handler.setLevel(logging.ERROR)
+    err_handler.setFormatter(formatter)
+    logger.addHandler(err_handler)
```
Can remove duplication by splitting `get_logger` into two methods like `create-logger(module)` which creates the logger, and then`logger.add_handler(create_handler(handler, log_level, formatter))`, where `create_handler` checks if `formatter` is None. The check is for the case when `AlertManager.get_alert_logger` is called, which doesn't have a formatter.

tttruong701 commented on src/collection.py:
```diff
@@ -0,0 +1,118 @@
+    separated into requests and responses.
+
+    Attributes:
+        sniffer (scapy.sendrecv.AsyncSniffer): packet sniffer, set to filter on port 80, and turn filter out received
```
I think you meant to remove `turn`? 

`and filter out received packets that are not HTTP related`.

tttruong701 commented on src/collection.py:
```diff
@@ -0,0 +1,118 @@
+    def stop_packet_collection(self):
+        """
+        :return: None
+        Stop the sniff.
```
🚫 👃

tttruong701 commented on src/collection.py:
```diff
@@ -0,0 +1,118 @@
+        if not self.sniffer:
+            logger.info(f'Starting packet listener with interface: {interface} listening for traffic on port 80.')
+            self.sniffer = AsyncSniffer(iface=interface, prn=self.process_packet, filter="tcp port 80")
+            self.sniffer.start()
```
👃

tttruong701 commented on src/alerting.py:
```diff
@@ -0,0 +1,89 @@
+        self.alert_logfile = '/var/log/past_alerts.log'
+        self.alert_logger = self.get_alert_logger()
+
+    def get_alert_logger(self):
```
Similar to my comment in `utils.py`, I think this should move and be encapsulated in the logging utility methods there.

tttruong701 commented on src/alerting.py:
```diff
@@ -0,0 +1,89 @@
+        if not self.past_alerts:
+            return 'No Past alerts to display.\n'
+        past_alerts_str = 'Past Alerts:\n'
+        if len(self.past_alerts) > 5:
```
Can replace 5 with a constant like self.threshold.

tttruong701 replied:
Can remove duplication by creating boolean for threshold condition, then use it to determine the collection of alerts to traverse, and then use it to determine if the extra log message must be logged.

Something like:
```
threshold = len(self.past_alerts) > 5
alerts = self.past_alerts[::-1][:5] if threshold else self.past_alerts[::-1]
for alert_start_time, alert_end_time in alerts:
  past_alerts_str += f'\tPast alert started: {alert_start_time}, ended: {alert_end_time}.\n'
if threshold
  extra_alert
```

tttruong701 commented on src/alerting.py:
```diff
@@ -0,0 +1,89 @@
+        if average_hits > self.alert_threshold and not self.alert_active:
+            logger.info(f'Average above threshold and no alert active. Starting new alert.')
+            self.alert_active = True
+            self.current_alert_start_time = strftime("%a, %d %b %Y %H:%M:%S", localtime())
```
Optional: Start timer (localtime()) at the beginning of the block for potential better accuracy.

tttruong701 commented on README.md:
```diff
@@ -1 +1,64 @@
-# http_sniffer
```
This is great!

tttruong701 commented on src/main.py:
```diff
@@ -0,0 +1,44 @@
+logger = utils.get_logger(__name__)
+
+def main(args):
+    interval_count = 0
+    statistics_check_interval_seconds = 10
+    alert_check_interval = 12
```
For configurability, maybe make these program arguments too with these values as default? Should update the README too if so.

tttruong701 commented on src/main.py:
```diff
@@ -0,0 +1,44 @@
+
+    while True:
+        sleep(statistics_check_interval_seconds)
+        interval_count += 1
```
Why not sleep for 12 seconds? Not sure what `interval_count` is used for besides being 0 or 1.

Probably easier to talk offline about this.

tttruong701 reviewed (changes requested):
Overall great! More comments

BadJukeBox committed:
Added testing, updated some files to make tests a little easier to understand.

BadJukeBox committed:
utils got skipped

BadJukeBox committed:
Fixed more test stuff

BadJukeBox committed:
Last documentation

BadJukeBox committed:
Changed some path stuff and added pythonpath to docker to resolve the issues. added exception handler to the main loop to handle CRTL+C
